internal/service/projectipaccesslist/model_project_ip_access_list_test.go

```diff
@@ -0,0 +1,215 @@
+package projectipaccesslist_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-framework-timeouts/resource/timeouts"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/conversion"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/service/projectipaccesslist"
+	"github.com/stretchr/testify/assert"
+	"go.mongodb.org/atlas-sdk/v20231115002/admin"
+)
+
+var (
+	AWSSecurityGroup = "AWSSecurityGroup"
+	CIDRBlock        = "CIDRBlock"
+	IPAddress        = "IPAddress"
+	Comment          = "Comment"
+	GroupID          = "GroupID"
+	IPAddressID      = conversion.EncodeStateID(map[string]string{
+		"entry":      IPAddress,
+		"project_id": GroupID,
+	})
+	CidrBlockID = conversion.EncodeStateID(map[string]string{
+		"entry":      CIDRBlock,
+		"project_id": GroupID,
+	})
+	AwsSecurityGroupID = conversion.EncodeStateID(map[string]string{
+		"entry":      AWSSecurityGroup,
+		"project_id": GroupID,
+	})
+)
+
+func TestNewMongoDBProjectIPAccessList(t *testing.T) {
+	testCases := []struct {
+		tfModel        *projectipaccesslist.TfProjectIPAccessListModel
+		expectedResult *[]admin.NetworkPermissionEntry
+		name           string
+	}{
+		{
+			name: "NewMongoDBProjectIPAccessList",
+			tfModel: &projectipaccesslist.TfProjectIPAccessListModel{
+				AWSSecurityGroup: types.StringValue(AWSSecurityGroup),
+				CIDRBlock:        types.StringValue(CIDRBlock),
+				IPAddress:        types.StringValue(IPAddress),
+				Comment:          types.StringValue(Comment),
+			},
+			expectedResult: &[]admin.NetworkPermissionEntry{
+				{
+					AwsSecurityGroup: &AWSSecurityGroup,
+					CidrBlock:        &CIDRBlock,
+					IpAddress:        &IPAddress,
+					Comment:          &Comment,
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			resultModel := projectipaccesslist.NewMongoDBProjectIPAccessList(tc.tfModel)
+
+			assert.Equal(t, tc.expectedResult, resultModel)
+		})
+	}
+}
+
+func TestNewTfProjectIPAccessListModel(t *testing.T) {
+	testCases := []struct {
+		tfModel        *projectipaccesslist.TfProjectIPAccessListModel
+		sdkModel       *admin.NetworkPermissionEntry
+		expectedResult *projectipaccesslist.TfProjectIPAccessListModel
+		name           string
+	}{
+		{
+			name: "NewTfProjectIPAccessListModel with IpAddress",
+			tfModel: &projectipaccesslist.TfProjectIPAccessListModel{
+				Timeouts: timeouts.Value{},
+			},
+			sdkModel: &admin.NetworkPermissionEntry{
+				IpAddress: &IPAddress,
+				Comment:   &Comment,
+				GroupId:   &GroupID,
+			},
+			expectedResult: &projectipaccesslist.TfProjectIPAccessListModel{
+				ID:               types.StringValue(IPAddressID),
+				ProjectID:        types.StringValue(GroupID),
+				AWSSecurityGroup: types.StringValue(""),
+				CIDRBlock:        types.StringValue(""),
```

Comment:
Interesting that we are storing empty strings instead of defining terraform null value. Don't believe we need to take immediate action here, but nice to see how unit testing evidences this behaviour.

Reply:
Yes its an interesting behaviour made explicit by the unit test. In this case, these attributes have `omitempty` (see [docs](https://pkg.go.dev/go.mongodb.org/atlas-sdk/v20231115002@v20231115002.1.0/admin#NetworkPermissionEntry)) so in this case when sent to the API they would be ignored (only one of `AWSSecurityGroup`, `CIDRBlock`, `IPAddress` can be set at the same time, as explained in the docs: `Your IP access list entry can be one **awsSecurityGroup**, one **cidrBlock**, or one **ipAddress**.` )

Reply:
actually given that these fields are pointers, the `omitempty` will work by including the property in the json if the pointer is not `nil` (even if it is a pointer to ""). In this case we are not including in the request because in https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/internal/service/projectipaccesslist/resource_project_ip_access_list.go#L237 we define the field as nil if state has an empty string.
TLDR: it is working well, but if we use proper terraform null values the state would be more "correct", and simplify handling these fields in model conversions.